Add support for PhyML interoperability

jmenglund opened the pull request:
I made a very simple wrapper for PhyML, which you might want to add to DendroPy. If you want tests I could try to write some, but I think I need some guidance to get started.

jeetsukumaran commented:
Hi Markus,

Thanks for this. Really useful, and much appreciated. 

I think tests would be great, but it is always a challenge to write tests for complex processes with non-deterministic outcomes. In this case, the three strategies I can think of would be: (a) check to make sure all parameters/arguments are being passed to the underlying PhyML correctly by inspection of the command list; (b) running on some simple data sets and making sure some aspects of the results match up; (c) feeding the results parser and known PhyML output and make sure it parses it correctly. 

In either case, at the very least, it would be good to have the examples run out of the box. Unfortunately, the current ones do not:

``phyml_estimate_tree.py`` results in:
~~~
Traceback (most recent call last):
  File "phyml_estimate_tree.py", line 9, in <module>
    result = phyml.run_phyml(
NameError: name 'phyml' is not defined
~~~

Fixing the issues (an import error), then results in:
~~~
[venv:python-3.5.0][examples:phyml1:3e76b61:+M+U]$ p phyml_estimate_tree.py 
Traceback (most recent call last):
  File "phyml_estimate_tree.py", line 13, in <module>
    gamma_cats=1)
  File "/Users/jeet/Documents/Projects/Evoinformatics/DendroPy/dendropy/dendropy/interop/phyml.py", line 256, in run_phyml
    char_matrix_f.name + "_phyml_tree", "newick")
  File "/Users/jeet/Documents/Projects/Evoinformatics/DendroPy/dendropy/dendropy/datamodel/basemodel.py", line 215, in get_from_path
    with open(src, "r", newline=None) as fsrc:
  File "/Users/jeet/Documents/Projects/Evoinformatics/DendroPy/dendropy/dendropy/utility/filesys.py", line 60, in pre_py34_open
    buffering=buffering)
IOError: [Errno 2] No such file or directory: '/var/folders/0y/qj2hjjl52rb4rwmfy984j25w0000gn/T/tmp7sdqfT_phyml_tree'
~~~

Similarly, ``phyml_estimate_model.py`` does not work out of the box either, due to different numbers of taxa in the tree and data file. Fixing this issue results in:
~~~
Traceback (most recent call last):
  File "phyml_estimate_tree.py", line 13, in <module>
    gamma_cats=1)
  File "/Users/jeet/Documents/Projects/Evoinformatics/DendroPy/dendropy/dendropy/interop/phyml.py", line 256, in run_phyml
    char_matrix_f.name + "_phyml_tree", "newick")
  File "/Users/jeet/Documents/Projects/Evoinformatics/DendroPy/dendropy/dendropy/datamodel/basemodel.py", line 215, in get_from_path
    with open(src, "r", newline=None) as fsrc:
  File "/Users/jeet/Documents/Projects/Evoinformatics/DendroPy/dendropy/dendropy/utility/filesys.py", line 60, in pre_py34_open
    buffering=buffering)
IOError: [Errno 2] No such file or directory: '/var/folders/0y/qj2hjjl52rb4rwmfy984j25w0000gn/T/tmp7sdqfT_phyml_tree'
~~~

I suspect the problem is the the temporary files are being removed too quickly. Why this is happening I'm not sure at the moment.

jmenglund commented:
Hi Jeet,

Sorry for this late response. Now I finally got some time to go through the issues you reported. I hope you'll get the code to work this time! I made the following fixes:

* PhyML output files are now allowed to have different name endings. It seems that some (older?) versions of PhyML add a ".txt" extension to the filenames. A filename mismatch could be the reason for your `IOError`.
* I fixed the two issues with the examples (import error and wrong user tree, respectively).

I also did some other changes. Here are the most important ones:

* I made `phyml_path` a required argument to `run_phyml` (I think it is better to be explicit about the path)
* I added the attribute `output_files` to `PhymlResult`. This attribute will show if any output file was missing when the result was collected.

Thanks for guidance on writing the tests. I will look into this once you get the code to work!
